src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -0,0 +1,165 @@
+
+use std::str::FromStr;
+
+use serde::de::Error;
+use serde::{Deserialize, Deserializer};
+
+/// Types common for template definitions for both
+/// Aarch64 and x86_64.
+
+/// Bit-mask based wrapper to apply
+/// changes to a given register's value
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub struct RegisterModifier {
+    /// Pointer of the location to be bit masked.
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    addr: u64,
+    /// Mask to be applied to register's value at the address
+    /// provided.
+    mask: BitMask,
+}
+
+/// Bit mask to adjust targeted bits to the value,
+/// applied by the filter only.
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub struct BitMask {
+    /// Mask value is to be applied according
+    /// to what is allowed through by the filter.
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    filter: u64,
+    /// Mask value to be applied.
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    value: u64,
+}
+
+/// Generic struct for addressing registers
+/// that are non-CPUID based.
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub struct RegisterId {
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    value: u64,
+}
+
+/// Guest config sub-module specifically useful for
+/// config templates.
+#[cfg(target_arch = "x86_64")]
+pub mod x86_64 {
+
+    use serde::Deserialize;
+
+    use crate::guest_config::templates::RegisterModifier;
+
+    /// CPUID register enumeration
+    #[allow(unused)]
+    #[allow(missing_docs)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub enum CpuidRegister {
+        Eax,
+        Ebx,
+        Ecx,
+        Edx,
+    }
+
+    /// Composite type that holistically provides
+    /// the location of a specific register being used
+    /// in the context of a CPUID tree.
+    #[allow(unused)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct CpuidRegisterId {
+        leaf: u32,
+        subleaf: u32,
+        flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        pointer: CpuidRegister,
+    }
+
+    /// Wrapper type to containing x86_64 CPU config modifiers.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct X86_64CpuTemplate {
+        /// Modifiers for CPUID configuration.
+        // pub cpuid_modifiers: Vec<RegisterBitMask>,
+        /// Modifiers for model specific registers.
+        pub msr_modifiers: Vec<RegisterModifier>,
+    }
+}
+
+/// Guest config sub-module specifically useful for
+/// config templates.
+#[cfg(target_arch = "aarch64")]
+pub mod aarch64 {}
+
+fn deserialize_u64_from_str<'de, D>(deserializer: D) -> Result<u64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let number_str = String::deserialize(deserializer)?;
+    let deserialized_number: u64 = if number_str.len() > 2 {
+        match &number_str[0..2] {
+            "0b" => u64::from_str_radix(&number_str[2..], 2),
+            "0x" => u64::from_str_radix(&number_str[2..], 16),
```

Comment:
what is the plan now? are we migrating to the tristate notation within this PR or separately?

Reply:
So the `deserialize_u64_from_str` in particular is for addresses, not the masks.

With regards to the tristate notation, I thought it was best to include it here. You can see an example of the tristate based template in the unit tests.